src/main/java/org/springframework/integration/kafka/inbound/KafkaMessageDrivenChannelAdapter.java

```diff
@@ -271,6 +288,32 @@ public int afterShutdown() {
 		return getPhase();
 	}
 
+	protected void setAttributesIfNecessary(Object record, Message<?> message) {
+		boolean needAttributes = getErrorChannel() != null
+				&& KafkaMessageDrivenChannelAdapter.this.retryTemplate == null;
+		if (needAttributes) {
+			attributesHolder.set(ErrorMessageUtils.getAttributeAccessor(null, null));
+		}
+		if (needAttributes || KafkaMessageDrivenChannelAdapter.this.recoveryCallback != null) {
+			AttributeAccessor attributes = attributesHolder.get();
+			if (attributes != null) {
+				attributes.setAttribute(ErrorMessageUtils.INPUT_MESSAGE_CONTEXT_KEY, message);
```

Comment:
As a compromise I think we should rely on the `RetryTemplate` if it is present, and `null` `errorChannel` and recommend to use `RecoveryCallback` which is channel publishing aware if that is a wish for such a error handling behavior.

In other words `errorChannel` and `retryTempalte` are mutually exclusive options. The `RecoveryCallback` callback can compensate `errorChannel` option.

On the other hand we always can populate internal `ErrorMessageSendingRecoverer` (if no by user) based on the provided `errorChannel` and with the `RawRecordHeaderErrorMessageStrategy` and propagate `null` for `errorChannel` to `super`. This way we really will have almost the same error handling scenario with and without `RetryTemplate`.

And that's maybe why I thought to have `ErrorMessagePublisher` in the `MPS` instead of just `EMS`. This way we would have only one place to build and send `ErrorMessage`...